libraries/shared/logs/extractor.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/vulcanize/vulcanizedb/pkg/core"
+	"github.com/vulcanize/vulcanizedb/pkg/datastore"
+)
+
+var ErrNoWatchedAddresses = errors.New("no watched addresses configured in the log extractor")
+
+type ILogExtractor interface {
+	AddTransformerConfig(config transformer.EventTransformerConfig)
+	ExtractLogs(recheckHeaders constants.TransformerExecution, errs chan error, missingHeadersFound chan bool)
+}
+
+type LogExtractor struct {
+	Addresses                []common.Address
+	CheckedHeadersRepository datastore.CheckedHeadersRepository
+	Fetcher                  fetcher.ILogFetcher
+	LogRepository            datastore.HeaderSyncLogRepository
+	StartingBlock            *int64
+	Syncer                   transactions.ITransactionsSyncer
+	Topics                   []common.Hash
+}
+
+// Add additional logs to extract
+func (extractor *LogExtractor) AddTransformerConfig(config transformer.EventTransformerConfig) {
+	if extractor.StartingBlock == nil {
+		extractor.StartingBlock = &config.StartingBlockNumber
+	} else if earlierStartingBlockNumber(config.StartingBlockNumber, *extractor.StartingBlock) {
+		extractor.StartingBlock = &config.StartingBlockNumber
+	}
+
+	addresses := transformer.HexStringsToAddresses(config.ContractAddresses)
+	extractor.Addresses = append(extractor.Addresses, addresses...)
+	extractor.Topics = append(extractor.Topics, common.HexToHash(config.Topic))
+}
+
+// Fetch and persist watched logs
+func (extractor LogExtractor) ExtractLogs(recheckHeaders constants.TransformerExecution, errs chan error, missingHeadersFound chan bool) {
+	if len(extractor.Addresses) < 1 {
+		logrus.Errorf("error extracting logs: %s", ErrNoWatchedAddresses.Error())
+		errs <- ErrNoWatchedAddresses
+		return
+	}
+
+	missingHeaders, missingHeadersErr := extractor.CheckedHeadersRepository.MissingHeaders(*extractor.StartingBlock, -1, getCheckCount(recheckHeaders))
+	if missingHeadersErr != nil {
+		logrus.Errorf("error fetching missing headers: %s", missingHeadersErr)
+		errs <- missingHeadersErr
+		return
+	}
+
+	if len(missingHeaders) < 1 {
+		missingHeadersFound <- false
+		return
+	}
+
+	for _, header := range missingHeaders {
+		logs, fetchLogsErr := extractor.Fetcher.FetchLogs(extractor.Addresses, extractor.Topics, header)
+		if fetchLogsErr != nil {
+			logError("error fetching logs for header: %s", fetchLogsErr, header)
+			errs <- fetchLogsErr
+			return
+		}
+
+		if len(logs) > 0 {
+			transactionsSyncErr := extractor.Syncer.SyncTransactions(header.Id, logs)
+			if transactionsSyncErr != nil {
+				logError("error syncing transactions: %s", transactionsSyncErr, header)
+				errs <- transactionsSyncErr
+				return
+			}
+
+			createLogsErr := extractor.LogRepository.CreateHeaderSyncLogs(header.Id, logs)
+			if createLogsErr != nil {
+				logError("error persisting logs: %s", createLogsErr, header)
+				errs <- createLogsErr
+				return
+			}
+		}
+
+		markHeaderCheckedErr := extractor.CheckedHeadersRepository.MarkHeaderChecked(header.Id)
```

Comment:
Good question! We do not record which headers have been checked for which logs. Would welcome any thoughts or ideas about adding new transformers on the fly...

Reply:
This PR then leads to a clear loss of certainty, doesn't it? I'm not 100% on how the recheck logic works, but it _has_ to be more expensive than keeping track of checked logs + the theoretical bloom filter check to verify emptiness no?

Thinking out loud about the current architecture (in this PR):
* When adding `newTransformer` to an already-synced vDB on restart, we get a new contract address and/or a new topic to care about.
* The `LogFetcher` needs to know that from the configured starting block of this transformer, we haven't yet checked for metadata relating to `newTransformer`.
* `LogFetcher` wont do anything about this unless it gets a header from `CheckedHeadersRepository`

So, given the current logic is based around missing headers, could we maybe set the headers from `startingBlockNumber -> head` as unchecked, and have the `logs` table as a cache to prevent us from asking the RPC nodes for logs we already have? A shortcut evaluation of the extractor basically. Some index of this could be loaded into memory perhaps.

What I'm saying is basically that if we manipulate the fact that a header is checked, and don't perform work for logs we already have, we may get rechecking for free-ish and it would also support adding new transformers to an already synced up instance.

(I'm not at all sure about this by the way :) )

Reply:
Two more things:
a) Just wanna be clear on that this is awesome and I love getting rid of the complexities around the checked headers table. :trophy: I _am_ a bit worried that we lose a quite important guarantee though.
b) If we find _one_ existence of `topic0` log in a certain block, we might not be able to use the bloom filter to make sure there aren't _more_ of that log : (

Reply:
Yeah this is a tricky one. I guess I wasn't thinking the same thing as you about the importance of being able to add a new transformer on the fly. Happy to put in some work toward making sure we preserve that.

I do think using `recheck` in that situation is going to be maximally expensive, since it means rechecking all headers for all logs.

I like the idea of using bloom filters as a short cut, but I'm having trouble wrapping my head around how it solves the problem without non-trivial additional complexity (e.g. specifically designating a new `topic0` as new and checking all checked headers' bloom filters for the possible presence of that value and querying for the log if so).

I suspect there's a similar issue with using the `logs` table as a cache. Unless we make strong assumptions about the order that logs were checked, it seems difficult to know that `checked_headers` were checked for a given log just based on the presence of a matching log in the `logs` table (since we theoretically could have checked only one block for that log).

I think marking all headers `startingBlockNumber -> head` as not checked would be a bridge to get there, but that approaches the same expense as `recheck` in the pessimistic case.

Thinking maybe we want to preserve `individual_log_checked` columns on the `checked_headers` table, but that should be more simple now that everything can be done in the extractor. Probably going to make this PR a bit bigger though :/